File: src/main/GO/leetcode/lists/hot100/LeetCode560.go
package hot100

func subarraySum(nums []int, k int) int {
	if len(nums) == 0 {
		return 0
	}
	mp := make(map[int]int)
	sum := make([]int, len(nums)+1)
	sum[0] = 0
	for i := 1; i <= len(nums); i++ {
		sum[i] = sum[i-1] + nums[i-1]
	}
	var res int
	for i := 0; i < len(sum); i++ {
		if v, ok := mp[sum[i]-k]; ok {
			res += v
		}
		mp[sum[i]] = mp[sum[i]] + 1
	}
	return res
}

File: src/main/java/leetcode/lists/hot100/LeetCode560.java
package leetcode.lists.hot100;

import java.util.HashMap;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/19 14:12
 */
public class LeetCode560 {

    public int subarraySum(int[] nums, int k) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int count = 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        int sum = 0;
        map.put(0, 1);
        for (int num : nums) {
            sum = sum + num;
            if (map.containsKey(sum - k)) {
                count = count + map.get(sum - k);
            }
            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }
        return count;
    }
}
